demos/create_workspace.ipynb

```diff
@@ -0,0 +1,164 @@
+    "from adam import Permission"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Edit this to reflect the user and parent workspace for which you would like to add a workspace.\n",
+    "\n",
+    "Note: it's unusual for dev and prod to have the same uuids for anything - they happen to have the same uuids for a few objects created early on in the life of ADAM because at one point we copied the table contents. For newly created objects uuids will be assigned independently."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "name = # Fill this in!\n",
+    "email = # Fill this in!\n",
+    "description = \"Workspace for \" + name + \" within Asteroid Institute\"\n",
+    "parent_project_prod = \"74020cfb-6528-4fb8-b170-5e7b2b0dc737\"  # The Asteroid Institute shared parent project.\n",
+    "parent_project_dev = \"74020cfb-6528-4fb8-b170-5e7b2b0dc737\"  # The Asteroid Institute shared parent project.\n",
+    "team_group_prod = \"ae62ec66-836c-4c4c-8e25-c643ec46b58c\"  # The Asteroid Institute team group.\n",
+    "team_group_dev = \"ae62ec66-836c-4c4c-8e25-c643ec46b58c\"  # The Asteroid Institute team group."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This sets up a workspace in prod for the given user."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(\"Setting up project and group membership for \" + name + \" in prod.\")\n",
+    "\n",
+    "config = ConfigManager(os.getcwd() + '/config.json').get_config('prod')\n",
+    "service = Service(config)\n",
+    "service.setup()\n",
+    "\n",
+    "# Adds the new project if there isn't one.\n",
+    "projects = service.get_projects_module()\n",
+    "project = None\n",
+    "for p in projects.get_sub_projects(parent_project_prod):\n",
+    "    if p.get_name() == name:\n",
+    "        project = p\n",
+    "        print(\"Found existing workspace for \" + name + \" with uuid \" + p.get_uuid())\n",
+    "if project is None:\n",
+    "    project = projects.new_project(parent_project_prod, name, description)\n",
+    "    print(\"Created workspace for \" + name + \" with uuid \" + project.get_uuid())\n",
+    "\n",
+    "    # Gives permission to the new project.\n",
+    "    permissions = service.get_permissions_module()\n",
+    "    permissions.grant_user_permission(email, Permission('ADMIN', 'PROJECT', project.get_uuid()))\n",
+    "    print(\"Granted admin permission to \" + email + \" on project \" + project.get_uuid())\n",
+    "\n",
+    "# Adds user to team group, if they're not already in it.\n",
+    "if team_group_prod is not None:\n",
+    "    groups = service.get_groups_module()\n",
+    "    if not email in [g.get_id() for g in groups.get_group_members(team_group_prod) if g.get_type() == \"USER\"]:\n",
```

Comment:
Hmmm... Is this addition idempotent? We were back and forth about this.
Also, you can probably get all direct groups of the user and check inclusion that way.

Reply:
I believe it is idempotent (and ran twice to see, seems to be fine), so we could avoid the check. That said, in this case I kinda like being able to report to the person running the notebook whether the new user was already part of the group. Any particular reason to check inclusion of group in user's groups rather than user in group's users? I set it up the other way so that if the person running is interested, it's easy to modify the code to list the group's users.